### PR TITLE
Add typing to Layer.__init__

### PR DESCRIPTION
**Pull request by dstansby**
Some more typing - there were also a couple of code fixes I had to make after adding the typing.

### CODE REVIEW

File: napari/types.py [hunk 1]
```diff
@@ -51,6 +52,11 @@
     'VectorsData',
     'LayerDataTuple',
     'image_reader_to_layerdata_reader',
+    'BlendMode',
```
Comment (psobolewskiPhD) [inline napari/types.py]:
```suggestion
    'BlendingMode',
```

File: napari/types.py [hunk 2]
```diff
@@ -51,6 +52,11 @@
...
+]
+
+BlendMode = Literal[
```
Comment (psobolewskiPhD) [inline napari/types.py]:
```suggestion
BlendingMode = Literal[
```

**Review by psobolewskiPhD — APPROVED** [review]
Small suggestion

File: napari/layers/base/base.py [hunk 3]
```diff
@@ -390,9 +395,9 @@ def __init__(
         # 4. `world2grid`: An additional transform mapping world-coordinates
         #   into a grid for looking at layers side-by-side.
         if scale is None:
-            scale = [1] * ndim
+            scale = tuple([1] * ndim)
```
Comment (Czaki) [inline napari/layers/base/base.py]:
```suggestion
            scale = (1,) * ndim
```

File: napari/layers/base/base.py [hunk 4]
```diff
@@ -390,9 +395,9 @@ def __init__(
...
         if translate is None:
-            translate = [0] * ndim
+            translate = tuple([0] * ndim)
```
Comment (Czaki) [inline napari/layers/base/base.py]:
```suggestion
            translate = (0,) * ndim
```

File: napari/layers/base/base.py [hunk 5]
```diff
@@ -306,25 +307,29 @@ class Layer(KeymapProvider, MousemapProvider, ABC, metaclass=PostInit):
 
     def __init__(
         self,
-        data,
-        ndim,
+        data: Union[npt.NDArray, list[npt.NDArray]],
```
Comment (Czaki) [inline napari/layers/base/base.py:310]: It could be also dask, zarr, tensorstore.

File: napari/layers/base/base.py [hunk 6]
```diff
@@ -306,25 +307,29 @@ class Layer(KeymapProvider, MousemapProvider, ABC, metaclass=PostInit):
...
-        name=None,
-        opacity=1.0,
-        projection_mode='none',
-        rotate=None,
-        scale=None,
-        shear=None,
-        translate=None,
-        visible=True,
-    ):
+        affine: Optional[Union[npt.NDArray, Affine]] = None,
```
Comment (Czaki) [inline napari/layers/base/base.py:313]:
```suggestion
        affine: Union[npt.NDArray, Affine, None] = None,
```

File: napari/layers/base/base.py [hunk 7]
```diff
@@ -306,25 +307,29 @@ class Layer(KeymapProvider, MousemapProvider, ABC, metaclass=PostInit):
...
+        blending: napari.types.BlendMode = 'translucent',
```
Comment (Czaki) [inline napari/layers/base/base.py]:
```suggestion
        blending: napari.types.BlendingMode = 'translucent',
```

File: napari/layers/base/base.py [hunk 8]
```diff
@@ -306,25 +307,29 @@ class Layer(KeymapProvider, MousemapProvider, ABC, metaclass=PostInit):
...
+        cache: bool = True,  # this should move to future "data source" object.
+        experimental_clipping_planes: Optional[
+            Union[list[dict], list[ClippingPlane], ClippingPlaneList]
+        ] = None,
```
Comment (Czaki) [inline napari/layers/base/base.py:318]:
```suggestion
        experimental_clipping_planes: Union[list[dict], list[ClippingPlane], ClippingPlaneList, None] = None,
```

File: napari/layers/base/base.py [hunk 9]
```diff
@@ -306,25 +307,29 @@ class Layer(KeymapProvider, MousemapProvider, ABC, metaclass=PostInit):
...
+        metadata: Optional[dict] = None,
+        mode: str = 'pan_zoom',
+        multiscale: bool = False,
+        name: Optional[str] = None,
+        opacity: float = 1.0,
+        projection_mode: str = 'none',
+        rotate: Optional[
+            Union[float, tuple[float, float, float], npt.NDArray]
+        ] = None,
+        scale: Optional[tuple[float, ...]] = None,
```
Comment (Czaki) [inline napari/layers/base/base.py:328]:
```suggestion
        scale: Optional[Sequence[float, ...]] = None,
```

File: napari/layers/base/base.py [hunk 10]
```diff
@@ -306,25 +307,29 @@ class Layer(KeymapProvider, MousemapProvider, ABC, metaclass=PostInit):
...
+        shear: Optional[npt.NDArray] = None,
+        translate: Optional[tuple[float, ...]] = None,
```
Comment (Czaki) [inline napari/layers/base/base.py:330]:
```suggestion
        translate: Optional[Sequence[float, ...]] = None,
```

File: napari/layers/utils/layer_utils.py [hunk 11]
```diff
@@ -596,7 +596,7 @@ def compute_multiscale_level_and_corners(
 
 
 def coerce_affine(
-    affine: Union[npt.ArrayLike, Affine],
+    affine: Optional[Union[npt.ArrayLike, Affine]],
```
Comment (Czaki) [inline napari/layers/utils/layer_utils.py:599]:
```suggestion
    affine: Union[npt.ArrayLike, Affine, None],
```

**Comment by GenevieveBuckley** [comment]
The mypy CI check is currently failing due to the new `napari/layers/base/_test_util_sample_layer.py` file introduced in the main branch:
```
 napari/layers/base/_test_util_sample_layer.py:48: error: Unused "type: ignore" comment  [unused-ignore]
Found 1 error in 1 file (checked 667 source files)
```

Two options to fix this:
1. Move this file into a `_tests` module (so mypy will ignore it), and update the import statement in `test_qaction_layer.py` to match. Eg: `napari/layers/base/_test_util_sample_layer.py` -> 'napari/layers/base/_tests/_test_util_sample_layer.py'. It's a test file, so I'm a bit confused about why it's living outside a test folder to begin with anyway.
2. OR, merge `main` into this PR branch, then remove the comment `# type: ignore [no-untyped-call]` from line 48 of `_test_util_sample_layer.py`

**Comment by dstansby** [comment]
I'm going to close this as it's very out of date

**Pull request closed without merging**